File: LinkedList.py
"""

Time Linked List copy of list of integers

"""

def make_node(val, next=None):
    return [val, next]

def get_val(node):
    return node[0]

def get_next(node):
    return node[1]

def set_val(node, val):
    node[0] = val

def set_next(node, next):
    node[1] = next

def copy_list(ll):
    if ll:
        s = make_node(get_val(ll))
        ptr = get_next(ll)
        curr = s
        while get_next(ptr):
            set_next(curr, make_node(get_val(ptr)))
            curr = get_next(curr)
            ptr = get_next(ptr)
        return s

def make_list(size):
    s = make_node(0)
    l = s
    for i in range(1, size):
        n = make_node(i)
        set_next(l, n)
        l = n
    return s

if __name__ == "__main__":
    from datetime import datetime
    
    list_sizes = [{'size':10 ** v, 'pow':v} for v in range(1, 7+1)]
    
    def calc_time(n):
        t1 = datetime.now()
        l = make_list(n)
        l2 = copy_list(l)
        t2 = datetime.now()
        return (t2 - t1).total_seconds()
    
    for size in list_sizes:
        n = size['size']
        times = [calc_time(n) for i in range(10)]
        average = sum(times) / len(times)
        print('Took %f seconds to allocate and copy list of size 10^%i' % (average, size['pow']))
        

File: LinkedList.cpp
#include<iostream>
#include<stdio.h>
#include<ctime>

using namespace std;

struct Node {
    int val;
    Node* next;
};

Node* make_list(int length) {
    Node* start = NULL;
    for (int i = length - 1; i >= 0; i--) {
        Node* next = start;
        start = new Node;
        start->val = i;
        start->next = next;
    }
    return start;
}

void delete_list(Node* start) {
    while (start != NULL) {
        Node* next = start->next;
        delete start;
        start = next;
    }
}

Node* copy_list(Node* start) {
    Node* s = new Node;
    s->val = start->val;
    Node* ptr = start->next;
    Node* curr = s;
    while (ptr->next != NULL) {
        Node* tmp = new Node;
        tmp->val = ptr->val;
        curr->next = tmp;
        curr = tmp;
        ptr = ptr->next;
    }
    return s;
}

int main () {
    int size = 10;
    for (int i = 1; i <= 7; i++) {
        clock_t begin = clock();
        
        Node* list = make_list(size);
        Node* copy = copy_list(list);
        
        clock_t end = clock();
        
        delete_list(list);
        delete_list(copy);
        
        float elapsed_secds = float( end - begin ) /  CLOCKS_PER_SEC;
        
        printf("Took %f seconds to copy a linked list of size 10^%i\n", elapsed_secds, i);
        
        size *= 10;
    }
}

